### ISSUE TITLE
Proposal: either.catching should be a boundary operator

### ISSUE THREAD

**Issue by lbialy** [softwaremill/ox#150]
One of the least pleasant surprises when working with direct style scala using Ox is that we're back to exception throwing and lying signatures. I found out the hard way that one of the nicest things one gets from monadic effects is that they capture exceptions pervasively so in 99% of cases if a function gets you an `IO[A]`, you *will* get an `IO[A]` and not an exception. This is not true with ox either combinators as either does not naturally capture (and hence why maybe the idea to have `Result[A]` that has a `Throwable` error channel is not that bad idea? in a different discussion we could debate if `Try` catching on all combinators is still bad, given that `IO` monads also do that? maybe Ox should have combinators for `Try`?). Either way (pun not intended), given that `either {}` block does not catch exceptions but returns an `Either[E, A]` there seems to be no nice, useful way to work with both structured `either` and exceptions in one go. Let me visualize this:

```scala
import ox.*, either.*
import scala.util.*
import java.io.IOException

def niceSafeFunc(): Either[IOException, String] = Right("something") 
def thisFuncCanThrow(str: String): String = throw Exception("lying signature")

def safeFunc: Either[Exception, Int] = either {
  val first = niceSafeFunc().ok()
  // do things
  val output = thisFuncCanThrow(first)
  // do stuff
  output.toInt // this can also throw and one can easily forget about it when used to IO monads always catching
}
``` 
this obviously crashes:
```
scala> safeFunc
java.lang.Exception: lying signature
  at rs$line$2$.thisFuncCanThrow(rs$line$2:5)
  at rs$line$2$.safeFunc(rs$line$2:10)
  ... 30 elided
```

because exceptions subvert the type of `safeFunc`. With current state of ox when one is facing such unsafe APIs one can only do:

```scala
import ox.*, either.*
import scala.util.*
import java.io.IOException

def niceSafeFunc(): Either[IOException, String] = Right("something") 
def thisFuncCanThrow(str: String): String = throw Exception("lying signature")

def safeFunc: Either[Throwable, Int] = catching {
  either {
    val first = niceSafeFunc().ok()
    // do things
    val output = thisFuncCanThrow(first)
    // do stuff
    output.toInt // this can also throw and one can easily forget about it when used to IO monads always catching
  }
}.flatten
``` 
which does the job at the cost of changing the return type of `safeFunc` to `Either[Throwable, Int]` but it's also quite redundant - I want to use nice boundary-based combinators, I want to be able to use `.ok()` but to do that I have to have a nested `either` scope. Proposal is to turn `catching` into a boundary operator too:

from this:
```scala
inline def catching[T](inline t: => T): Either[Throwable, T] =
  try Right(t)
  catch case NonFatal(e) => Left(e)
```
 
to this:
 
 ```scala
inline def catching[T](inline t: Label[Either[Throwable, T]] ?=> T): Either[Throwable, T] =
  try boundary(Right(t))
  catch case NonFatal(e) => Left(e)
```

so that in the end we get:

```scala
import ox.*, either.*
import scala.util.*
import java.io.IOException

def niceSafeFunc(): Either[IOException, String] = Right("something") 
def thisFuncCanThrow(str: String): String = throw Exception("lying signature")

def safeFunc: Either[Throwable, Int] = either.catching {
  val first = niceSafeFunc().ok()
  // do things
  val output = thisFuncCanThrow(first)
  // do stuff
  output.toInt // this can also throw and one can easily forget about it when used to IO monads always catching
}
```

Your thoughts?

**Comment by lbialy** [comment]
Alternative is to have and use two control structures, as stdlib has it right now - `Try` and `Either`, a functor and bifunctor, and allow users to use whatever fits their needs the most? Given their differences in implementation, maybe a `Result` that doesn't catch on combinators but is just an either with `Throwable` channel instead of generic `Left` would fit better than `Try`? I actually doubt that - monadic effects went from "your monad is shitty if it captures exceptions" to "yeah, we capture exceptions on all IO combinators" and this seems to be the correct, unsurprising way of doing things in the end.

**Comment by lbialy** [comment]
This is also an option:
```scala
  import scala.util.{Try, Failure, Success}
  
  extension [A](inline t: Try[A])
    @targetName("ok_try")
    transparent inline def ok(): A =
      summonFrom {
        case given boundary.Label[Either[Throwable, Nothing]] =>
          t match
            case Failure(t) => break(Left(t))
            case Success(a) => a
        case given boundary.Label[Either[Nothing, Nothing]] =>
          error(
            "The enclosing `either` call uses a different error type.\nIf it's explicitly typed, is the error type correct? For `Try` error type has to be `Throwable`."
          )
        case _ => error("`.ok()` can only be used within an `either` call.\nIs it present?")
      }
 ```
  
 ```scala
    import ox.*, either.*
    import scala.util.*
    import java.io.IOException
    case class CaseEx(msg: String) extends Exception(msg)

    def niceSafeFunc(): Try[String] = Try("something")

    def thisFuncCanThrow(str: String): String = throw CaseEx("lying signature")

    def safeFunc: Either[Throwable, Int] = either[Throwable, Int] {
      val first = niceSafeFunc().ok()
      // do things
      val output: String = Try(thisFuncCanThrow(first)).ok()
      // do stuff
      output.toInt // this can also throw and one can easily forget about it when used to IO monads always catching
    }

    safeFunc shouldEqual Left(CaseEx("lying signature"))
```

**Comment by lbialy** [comment]
Haha, before we get into a very deep discussion about error channels, modeling of error channels (only exceptions are built-in, Try or Either require special handling and syntax) and whether exceptions should be used (I'm firmly in the *no, they shouldn't* team btw), let's get this out of the way:

Proposal is to turn catching into a boundary operator too:

from this:
```scala
inline def catching[T](inline t: => T): Either[Throwable, T] =
  try Right(t)
  catch case NonFatal(e) => Left(e)
```
to this:
```scala
inline def catching[T](inline t: Label[Either[Throwable, T]] ?=> T): Either[Throwable, T] =
 try boundary(Right(t))
 catch case NonFatal(e) => Left(e)
```

**Comment by adamw** [comment]
Exceptions are a fact of life on the JVM, I think the best you can do is try to push them to the role of "panics". But back to the main topic: I suppose, but then it would make sense to move it to `object either`. So that it's next to `apply` (the main boundary method), but it *also* catches exceptions & is specialised to throwables

**Comment by lbialy** [comment]
Sure, in Scala code - yeah, treating them as panics is probably mostly fine. Issue is: they are very often used for handling of things that should not be panics. A common example is, of course, `"not a number".toInt` which throws `NumberFormatException` because why not. If you have a piece of foreign code that does indeed do this and you want to handle that gracefully you have to catch and that's where `catching` is very helpful. It would just remove some boilerplate if it was a boundary for `Label[Either[Throwable, A]]`.

**Comment by adamw** [comment]
using .toInt instead of .toIntOption when you're not 100% sure that it's a number is a bug - especially when you're not catching the exception. But sure, let's add the label, and move it to `object either`

**Comment by lbialy** [comment]
Sure it is but that code can be in a dependency lib code. I'll push a PR
tomorrow.

**Comment by lbialy** [comment]
https://github.com/softwaremill/ox/pull/167

**Comment by lbialy** [comment]
implemented with #167
